CodeEdit/Features/Documents/WorkspaceDocument+Search.swift

```diff
@@ -9,106 +9,225 @@ import Foundation
+        /// That means if a search result is a file, and the search term appears in the file,
+        /// the function will add the line number, line content, and keyword range to the `SearchResultsViewModel`.
+        ///
+        /// - Parameters:
+        ///   - query: The search query string.
+        ///   - searchResults: An inout parameter containing the array of `SearchResultsViewModel` to be evaluated.
+        ///   It will be modified to include line matches.
+        private func evaluateResult(query: String, searchResult: inout SearchResultModel) async {
+            let searchResultCopy = searchResult
+            var newMatches = [SearchResultMatchModel]()
+
+            guard let data = try? Data(contentsOf: searchResult.file.url),
+                  let string = String(data: data, encoding: .utf8) else {
+                return
+            }
+
+            await withTaskGroup(of: SearchResultMatchModel?.self) { group in
+                for (lineNumber, line) in string.split(separator: "\n").lazy.enumerated() {
+                    group.addTask {
+                        let rawNoSapceLine = line.trimmingCharacters(in: .whitespacesAndNewlines)
+                        let noSpaceLine = rawNoSapceLine.lowercased()
+                        if self.lineContainsSearchTerm(line: noSpaceLine, term: query) {
+                            let matches = noSpaceLine.ranges(of: query).map { range in
+                                return [lineNumber, noSpaceLine, range]
+                            }
+
+                            for match in matches {
+                                if let lineNumber = match[0] as? Int,
+                                   let lineContent = match[1] as? String,
+                                   let keywordRange = match[2] as? Range<String.Index> {
+                                    let matchModel = SearchResultMatchModel(
+                                        lineNumber: lineNumber,
+                                        file: searchResultCopy.file,
+                                        lineContent: lineContent,
+                                        keywordRange: keywordRange
+                                    )
 
-                // If `fileSearchResult` isn't nil it means we've found matches in the file
-                // so we add it to the search results.
-                if let fileSearchResult {
-                    searchResult.append(fileSearchResult)
+                                    return matchModel
+                                }
+                            }
+                        }
+                        return nil
+                    }
+                    for await groupRes in group {
+                        if let groupRes {
+                            newMatches.append(groupRes)
+                        }
+                    }
                 }
             }
+            searchResult.lineMatches = newMatches
```

Comment:
One bug I found, otherwise this looks perfect. In this function, SearchKit will find matches that don't perfectly match the search term. In the following screenshots you can see SearchKit is matching `hello` but the search term `hello ` is not being found when scanning the file.

Results for `hello `:
<img width="327" alt="Screenshot 2023-11-29 at 10 38 21 AM" src="https://github.com/CodeEditApp/CodeEdit/assets/35942988/74e3c98f-a103-4723-af1c-d0a33b856ccf">

Results for `hello`:
<img width="332" alt="Screenshot 2023-11-29 at 10 38 29 AM" src="https://github.com/CodeEditApp/CodeEdit/assets/35942988/cfbdb87b-6503-4ff2-a39c-d6f5b1785c91">

One quick solution would be to make this function return a `bool` indicating if the searched file has any results, and if there are no results the file is ignored by the caller.

Reply:
Seems perfect to me!